Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,344 @@
+            {
+                var branch = repo.Branches[branchName];
+                if (branch == null)
+                {
+                    branch = repo.CreateBranch(branchName, startPoint);
+                    repo.Checkout(branch);
+                }
+                else if (string.IsNullOrEmpty(startPoint))
+                {
+                    repo.Checkout(branch);
+                }
+                else
+                {
+                    var commit = repo.Commits.Single(c => c.Sha == startPoint);
+                    if (commit != null)
+                    {
+                        repo.Checkout(branch);
+                        repo.Reset(ResetMode.Hard, commit);
+                    }
+                }
+            }
+        }
+
+        public override void UpdateRef(string toBeUpdatedToAlias)
+        {
+            UpdateRawRef(string.Format("refs/heads/{0}", "master"), string.Format("refs/heads/{0}", toBeUpdatedToAlias));
+        }
+
+        public void UpdateRawRef(string toBeUpdatedRef, string toBeUpdatedToRef)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Refs.UpdateTarget(toBeUpdatedRef, toBeUpdatedToRef);
+            }
+        }
+
+        public override bool DoesBranchContainCommit(string branchName, string commitOrBranchName)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var branch = repo.Branches[branchName];
+                var otherBranch = repo.Branches[commitOrBranchName];
+                if (branch == null) return false;
+                return repo.Commits.QueryBy(new CommitFilter { Since = branch }).Where(c => c.Sha == commitOrBranchName).Any() ||
+                       otherBranch != null
+                       ? repo.Commits.QueryBy(new CommitFilter { Since = branch }).Where(c => c.Sha == otherBranch.Tip.Sha).Any()
+                       : false;
+            }
+        }
+
+        public override IEnumerable<string> ListFiles(string path, SearchOption searchOption, params string[] lookupList)
+        {
+            path = PathUtility.CleanPath(path);
+
+            if (!FileSystemHelpers.IsSubfolder(RepositoryPath, path))
+            {
+                throw new NotSupportedException("Only paths relative to the repository root path are supported, path provided: '{0}' is not a child folder of '{1}'".FormatCurrentCulture(path, RepositoryPath));
+            }
+
+            if (!Directory.Exists(path)) return Enumerable.Empty<string>();
+
+            var lookupPatterns = lookupList.Select(l => string.Format("^{0}$", Regex.Escape(l).Replace("\\*", ".*")));
+
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var files = repo.Diff.Compare<TreeChanges>(null, DiffTargets.Index, lookupList, compareOptions: new CompareOptions(){IncludeUnmodified = true})
```

Comment:
I wasn't sure what it would do in this case considering the tree I am diffing against is an empty tree.